File: .github/agents/angular-design-system-analyzer/scripts/analyze_styles.py
#!/usr/bin/env python3
"""
Analyze CSS/SCSS styles in an Angular design system.
Extracts variables, theming system, and PrimeNG customizations.
"""

import os
import json
import sys
import re
from pathlib import Path
from typing import Dict, List, Any, Set


def analyze_styles(repo_path: str) -> Dict[str, Any]:
    """
    Analyze all style files in the repository.

    Args:
        repo_path: Path to the Angular repository

    Returns:
        Dictionary containing style analysis
    """
    repo_path = Path(repo_path).resolve()

    if not repo_path.exists():
        raise ValueError(f"Repository path does not exist: {repo_path}")

    analysis = {
        "global_styles": [],
        "component_styles": [],
        "css_variables": {},
        "scss_variables": {},
        "color_palette": {},
        "primeng_overrides": [],
        "theme_files": [],
        "mixins": [],
        "functions": [],
    }

    src_path = repo_path / "src"
    if not src_path.exists():
        raise ValueError(f"src directory not found in {repo_path}")

    # Analyze all style files
    for ext in ["*.css", "*.scss", "*.sass", "*.less"]:
        for style_file in src_path.rglob(ext):
            relative_path = style_file.relative_to(repo_path)

            # Classify as global or component style
            is_component_style = ".component." in style_file.name

            style_info = {
                "name": style_file.name,
                "path": str(relative_path),
                "type": style_file.suffix[1:],
                "is_component_style": is_component_style,
            }

            # Read and analyze content
            try:
                content = style_file.read_text(encoding="utf-8")

                # Extract variables
                if style_file.suffix in [".scss", ".sass"]:
                    scss_vars = extract_scss_variables(content)
                    if scss_vars:
                        style_info["variables"] = scss_vars
                        analysis["scss_variables"].update(scss_vars)

                    # Extract mixins and functions
                    mixins = extract_scss_mixins(content)
                    if mixins:
                        style_info["mixins"] = mixins
                        analysis["mixins"].extend(mixins)

                    functions = extract_scss_functions(content)
                    if functions:
                        style_info["functions"] = functions
                        analysis["functions"].extend(functions)

                # Extract CSS custom properties (variables)
                css_vars = extract_css_variables(content)
                if css_vars:
                    style_info["css_variables"] = css_vars
                    analysis["css_variables"].update(css_vars)

                # Detect PrimeNG overrides
                primeng_overrides = detect_primeng_overrides(content)
                if primeng_overrides:
                    style_info["primeng_overrides"] = primeng_overrides
                    analysis["primeng_overrides"].append(
                        {"file": str(relative_path), "overrides": primeng_overrides}
                    )

                # Detect theme files
                if is_theme_file(style_file.name, content):
                    analysis["theme_files"].append(style_info)

            except Exception as e:
                print(f"Warning: Could not read {style_file}: {e}")
                continue

            if is_component_style:
                analysis["component_styles"].append(style_info)
            else:
                analysis["global_styles"].append(style_info)

    # Extract color palette from variables
    analysis["color_palette"] = extract_color_palette(
        analysis["scss_variables"], analysis["css_variables"]
    )

    # Add statistics
    analysis["statistics"] = {
        "total_style_files": len(analysis["global_styles"])
        + len(analysis["component_styles"]),
        "global_styles": len(analysis["global_styles"]),
        "component_styles": len(analysis["component_styles"]),
        "scss_variables": len(analysis["scss_variables"]),
        "css_variables": len(analysis["css_variables"]),
        "colors_in_palette": len(analysis["color_palette"]),
        "primeng_override_files": len(analysis["primeng_overrides"]),
        "theme_files": len(analysis["theme_files"]),
        "mixins": len(analysis["mixins"]),
        "functions": len(analysis["functions"]),
    }

    return analysis


def extract_scss_variables(content: str) -> Dict[str, str]:
    """Extract SCSS variables ($variable: value;)."""
    variables = {}
    pattern = r"\$([a-zA-Z0-9_-]+)\s*:\s*([^;]+);"
    matches = re.findall(pattern, content)

    for var_name, var_value in matches:
        variables[var_name] = var_value.strip()

    return variables


def extract_css_variables(content: str) -> Dict[str, str]:
    """Extract CSS custom properties (--variable: value;)."""
    variables = {}
    pattern = r"--([a-zA-Z0-9_-]+)\s*:\s*([^;]+);"
    matches = re.findall(pattern, content)

    for var_name, var_value in matches:
        variables[var_name] = var_value.strip()

    return variables


def extract_scss_mixins(content: str) -> List[Dict[str, str]]:
    """Extract SCSS mixins."""
    mixins = []
    pattern = r"@mixin\s+([a-zA-Z0-9_-]+)\s*(\([^)]*\))?\s*{"
    matches = re.findall(pattern, content)

    for mixin_name, params in matches:
        mixins.append(
            {"name": mixin_name, "parameters": params.strip() if params else ""}
        )

    return mixins


def extract_scss_functions(content: str) -> List[Dict[str, str]]:
    """Extract SCSS functions."""
    functions = []
    pattern = r"@function\s+([a-zA-Z0-9_-]+)\s*(\([^)]*\))\s*{"
    matches = re.findall(pattern, content)

    for func_name, params in matches:
        functions.append({"name": func_name, "parameters": params.strip()})

    return functions


def detect_primeng_overrides(content: str) -> List[str]:
    """
    Detect PrimeNG component style overrides.

    Args:
        content: Style file content

    Returns:
        List of PrimeNG selectors being overridden
    """
    overrides = []

    # Common PrimeNG class patterns
    primeng_patterns = [
        r"\.p-[a-z-]+",  # PrimeNG classes start with .p-
        r"::ng-deep\s+\.p-[a-z-]+",  # Deep selectors
        r":host\s+::ng-deep\s+\.p-[a-z-]+",
    ]

    for pattern in primeng_patterns:
        matches = re.findall(pattern, content)
        overrides.extend(matches)

    # Remove duplicates and return
    return list(set(overrides))


def is_theme_file(filename: str, content: str) -> bool:
    """
    Determine if a file is a theme file.

    Args:
        filename: Name of the file
        content: File content

    Returns:
        True if file appears to be a theme file
    """
    # Check filename
    theme_keywords = ["theme", "variables", "colors", "palette"]
    if any(keyword in filename.lower() for keyword in theme_keywords):
        return True

    # Check content for theme-related patterns
    # Many color variables or CSS custom properties
    color_pattern = r"(#[0-9a-fA-F]{3,6}|rgb|hsl|var\(--)"
    color_matches = re.findall(color_pattern, content)

    # If file has many color definitions, likely a theme file
    return len(color_matches) > 10


def extract_color_palette(
    scss_vars: Dict[str, str], css_vars: Dict[str, str]
) -> Dict[str, str]:
    """
    Extract color palette from variables.

    Args:
        scss_vars: SCSS variables
        css_vars: CSS custom properties

    Returns:
        Dictionary of color variables
    """
    color_palette = {}

    # Color patterns
    color_pattern = r"^(#[0-9a-fA-F]{3,6}|rgb|rgba|hsl|hsla)"

    # Check SCSS variables
    for var_name, var_value in scss_vars.items():
        if re.match(color_pattern, var_value.strip()) or "color" in var_name.lower():
            color_palette[f"${var_name}"] = var_value

    # Check CSS variables
    for var_name, var_value in css_vars.items():
        if re.match(color_pattern, var_value.strip()) or "color" in var_name.lower():
            color_palette[f"--{var_name}"] = var_value

    return color_palette


def main():
    if len(sys.argv) < 2:
        print("Usage: python analyze_styles.py <repo_path> [output_file]")
        print("Example: python analyze_styles.py /path/to/angular-project styles.json")
        sys.exit(1)

    repo_path = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else "styles_analysis.json"

    try:
        print(f"Analyzing styles in: {repo_path}")
        analysis = analyze_styles(repo_path)

        # Write to output file
        with open(output_file, "w", encoding="utf-8") as f:
            json.dump(analysis, f, indent=2, ensure_ascii=False)

        print(f"\n✅ Analysis complete! Results saved to: {output_file}")
        print(f"\nStatistics:")
        for key, value in analysis["statistics"].items():
            print(f"  {key}: {value}")

        if analysis["color_palette"]:
            print(f"\nSample colors from palette:")
            for i, (var_name, var_value) in enumerate(
                list(analysis["color_palette"].items())[:5]
            ):
                print(f"  {var_name}: {var_value}")
            if len(analysis["color_palette"]) > 5:
                print(f"  ... and {len(analysis['color_palette']) - 5} more")

    except Exception as e:
        print(f"❌ Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: .github/agents/angular-design-system-analyzer-v2/scripts/analyze_styles.mjs
#!/usr/bin/env node

/**
 * Analyze CSS/SCSS styles in an Angular design system.
 * Extracts variables, theming system, and PrimeNG customizations.
 */

import fs from 'node:fs';
import path from 'node:path';

/**
 * Recursively find files matching given extensions.
 * @param {string} dir - Directory to search
 * @param {string[]} exts - File extensions to match (e.g. ['.css', '.scss'])
 * @returns {string[]} Array of matching file paths
 */
function rglobMulti(dir, exts) {
  const results = [];
  const skipDirs = new Set(['node_modules', 'dist', 'coverage', '.git', '.angular']);

  function walk(currentDir) {
    let entries;
    try {
      entries = fs.readdirSync(currentDir, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries) {
      if (entry.name.startsWith('.') && entry.name !== '.') continue;
      if (skipDirs.has(entry.name) && entry.isDirectory()) continue;

      const fullPath = path.join(currentDir, entry.name);
      if (entry.isDirectory()) {
        walk(fullPath);
      } else if (entry.isFile() && exts.some((ext) => entry.name.endsWith(ext))) {
        results.push(fullPath);
      }
    }
  }

  walk(dir);
  return results;
}

/**
 * Analyze all style files in the repository.
 * @param {string} repoPath - Path to the Angular repository
 * @returns {object} Style analysis
 */
function analyzeStyles(repoPath) {
  const resolvedPath = path.resolve(repoPath);

  if (!fs.existsSync(resolvedPath)) {
    throw new Error(`Repository path does not exist: ${resolvedPath}`);
  }

  const analysis = {
    global_styles: [],
    component_styles: [],
    css_variables: {},
    scss_variables: {},
    color_palette: {},
    primeng_overrides: [],
    theme_files: [],
    mixins: [],
    functions: [],
  };

  const srcPath = path.join(resolvedPath, 'src');
  if (!fs.existsSync(srcPath)) {
    throw new Error(`src directory not found in ${resolvedPath}`);
  }

  // Analyze all style files
  const styleFiles = rglobMulti(srcPath, ['.css', '.scss', '.sass', '.less']);

  for (const styleFile of styleFiles) {
    const relativePath = path.relative(resolvedPath, styleFile);
    const fileName = path.basename(styleFile);
    const ext = path.extname(styleFile);

    // Classify as global or component style
    const isComponentStyle = fileName.includes('.component.');

    const styleInfo = {
      name: fileName,
      path: relativePath,
      type: ext.slice(1),
      is_component_style: isComponentStyle,
    };

    // Read and analyze content
    let content;
    try {
      content = fs.readFileSync(styleFile, 'utf-8');
    } catch (e) {
      console.warn(`Warning: Could not read ${styleFile}: ${e.message}`);
      continue;
    }

    // Extract variables
    if (ext === '.scss' || ext === '.sass') {
      const scssVars = extractScssVariables(content);
      if (Object.keys(scssVars).length > 0) {
        styleInfo.variables = scssVars;
        Object.assign(analysis.scss_variables, scssVars);
      }

      // Extract mixins and functions
      const mixins = extractScssMixins(content);
      if (mixins.length > 0) {
        styleInfo.mixins = mixins;
        analysis.mixins.push(...mixins);
      }

      const functions = extractScssFunctions(content);
      if (functions.length > 0) {
        styleInfo.functions = functions;
        analysis.functions.push(...functions);
      }
    }

    // Extract CSS custom properties (variables)
    const cssVars = extractCssVariables(content);
    if (Object.keys(cssVars).length > 0) {
      styleInfo.css_variables = cssVars;
      Object.assign(analysis.css_variables, cssVars);
    }

    // Detect PrimeNG overrides
    const primengOverrides = detectPrimengOverrides(content);
    if (primengOverrides.length > 0) {
      styleInfo.primeng_overrides = primengOverrides;
      analysis.primeng_overrides.push({
        file: relativePath,
        overrides: primengOverrides,
      });
    }

    // Detect theme files
    if (isThemeFile(fileName, content)) {
      analysis.theme_files.push(styleInfo);
    }

    if (isComponentStyle) {
      analysis.component_styles.push(styleInfo);
    } else {
      analysis.global_styles.push(styleInfo);
    }
  }

  // Extract color palette from variables
  analysis.color_palette = extractColorPalette(analysis.scss_variables, analysis.css_variables);

  // Add statistics
  analysis.statistics = {
    total_style_files: analysis.global_styles.length + analysis.component_styles.length,
    global_styles: analysis.global_styles.length,
    component_styles: analysis.component_styles.length,
    scss_variables: Object.keys(analysis.scss_variables).length,
    css_variables: Object.keys(analysis.css_variables).length,
    colors_in_palette: Object.keys(analysis.color_palette).length,
    primeng_override_files: analysis.primeng_overrides.length,
    theme_files: analysis.theme_files.length,
    mixins: analysis.mixins.length,
    functions: analysis.functions.length,
  };

  return analysis;
}

/**
 * Extract SCSS variables ($variable: value;).
 * @param {string} content - Style file content
 * @returns {object} Map of variable names to values
 */
function extractScssVariables(content) {
  const variables = {};
  const pattern = /\$([a-zA-Z0-9_-]+)\s*:\s*([^;]+);/g;
  let match;
  while ((match = pattern.exec(content)) !== null) {
    variables[match[1]] = match[2].trim();
  }
  return variables;
}

/**
 * Extract CSS custom properties (--variable: value;).
 * @param {string} content - Style file content
 * @returns {object} Map of variable names to values
 */
function extractCssVariables(content) {
  const variables = {};
  const pattern = /--([a-zA-Z0-9_-]+)\s*:\s*([^;]+);/g;
  let match;
  while ((match = pattern.exec(content)) !== null) {
    variables[match[1]] = match[2].trim();
  }
  return variables;
}

/**
 * Extract SCSS mixins.
 * @param {string} content - Style file content
 * @returns {object[]}
 */
function extractScssMixins(content) {
  const mixins = [];
  const pattern = /@mixin\s+([a-zA-Z0-9_-]+)\s*(\([^)]*\))?\s*\{/g;
  let match;
  while ((match = pattern.exec(content)) !== null) {
    mixins.push({
      name: match[1],
      parameters: match[2] ? match[2].trim() : '',
    });
  }
  return mixins;
}

/**
 * Extract SCSS functions.
 * @param {string} content - Style file content
 * @returns {object[]}
 */
function extractScssFunctions(content) {
  const functions = [];
  const pattern = /@function\s+([a-zA-Z0-9_-]+)\s*(\([^)]*\))\s*\{/g;
  let match;
  while ((match = pattern.exec(content)) !== null) {
    functions.push({
      name: match[1],
      parameters: match[2].trim(),
    });
  }
  return functions;
}

/**
 * Detect PrimeNG component style overrides.
 * @param {string} content - Style file content
 * @returns {string[]} PrimeNG selectors being overridden
 */
function detectPrimengOverrides(content) {
  const overrides = new Set();

  const primengPatterns = [
    /\.p-[a-z-]+/g, // PrimeNG classes start with .p-
    /::ng-deep\s+\.p-[a-z-]+/g, // Deep selectors
    /:host\s+::ng-deep\s+\.p-[a-z-]+/g,
  ];

  for (const pattern of primengPatterns) {
    let match;
    while ((match = pattern.exec(content)) !== null) {
      overrides.add(match[0]);
    }
  }

  return [...overrides];
}

/**
 * Determine if a file is a theme file.
 * @param {string} filename - Name of the file
 * @param {string} content - File content
 * @returns {boolean}
 */
function isThemeFile(filename, content) {
  // Check filename
  const themeKeywords = ['theme', 'variables', 'colors', 'palette'];
  if (themeKeywords.some((kw) => filename.toLowerCase().includes(kw))) {
    return true;
  }

  // Check content for theme-related patterns
  const colorPattern = /(#[0-9a-fA-F]{3,6}|rgb|hsl|var\(--)/g;
  const colorMatches = content.match(colorPattern) || [];

  // If file has many color definitions, likely a theme file
  return colorMatches.length > 10;
}

/**
 * Extract color palette from variables.
 * @param {object} scssVars - SCSS variables
 * @param {object} cssVars - CSS custom properties
 * @returns {object} Color palette
 */
function extractColorPalette(scssVars, cssVars) {
  const colorPalette = {};
  const colorPattern = /^(#[0-9a-fA-F]{3,6}|rgb|rgba|hsl|hsla)/;

  // Check SCSS variables
  for (const [varName, varValue] of Object.entries(scssVars)) {
    if (colorPattern.test(varValue.trim()) || varName.toLowerCase().includes('color')) {
      colorPalette[`$${varName}`] = varValue;
    }
  }

  // Check CSS variables
  for (const [varName, varValue] of Object.entries(cssVars)) {
    if (colorPattern.test(varValue.trim()) || varName.toLowerCase().includes('color')) {
      colorPalette[`--${varName}`] = varValue;
    }
  }

  return colorPalette;
}

// --- Main ---
function main() {
  const args = process.argv.slice(2);

  if (args.length < 1) {
    console.log('Usage: node analyze_styles.mjs <repo_path> [output_file]');
    console.log('Example: node analyze_styles.mjs /path/to/angular-project styles.json');
    process.exit(1);
  }

  const repoPath = args[0];
  const outputFile = args[1] || 'styles_analysis.json';

  try {
    console.log(`Analyzing styles in: ${repoPath}`);
    const analysis = analyzeStyles(repoPath);

    // Write to output file
    fs.writeFileSync(outputFile, JSON.stringify(analysis, null, 2), 'utf-8');

    console.log(`\n✅ Analysis complete! Results saved to: ${outputFile}`);
    console.log('\nStatistics:');
    for (const [key, value] of Object.entries(analysis.statistics)) {
      console.log(`  ${key}: ${value}`);
    }

    const palette = Object.entries(analysis.color_palette);
    if (palette.length > 0) {
      console.log('\nSample colors from palette:');
      for (const [varName, varValue] of palette.slice(0, 5)) {
        console.log(`  ${varName}: ${varValue}`);
      }
      if (palette.length > 5) {
        console.log(`  ... and ${palette.length - 5} more`);
      }
    }
  } catch (e) {
    console.error(`❌ Error: ${e.message}`);
    process.exit(1);
  }
}

main();
